dali/pipeline/data/tensor_vector.cc

```diff
@@ -0,0 +1,439 @@
+  }
+}
+
+
+template <typename Backend>
+bool TensorVector<Backend>::is_pinned() const {
+  if (state_ == State::contiguous) {
+    return tl_->is_pinned();
+  }
+  if (curr_tensors_size_ == 0) {
+    return pinned_;
+  }
+  for (size_t i = 1; i < curr_tensors_size_; i++) {
+    assert(tensors_[i]->is_pinned() == tensors_[0]->is_pinned());
+  }
+  return tensors_[0]->is_pinned();
+}
+
+
+template <typename Backend>
+void TensorVector<Backend>::reserve(size_t total_bytes) {
+  state_ = State::contiguous;
+  tl_->reserve(total_bytes);
+}
+
+
+template <typename Backend>
+void TensorVector<Backend>::reserve(size_t bytes_per_sample, int batch_size) {
+  assert(batch_size > 0);
+  state_ = State::noncontiguous;
+  resize_tensors(batch_size);
+  for (auto t : tensors_) {
+    t->reserve(bytes_per_sample);
+  }
+}
+
+
+template <typename Backend>
+bool TensorVector<Backend>::IsContiguous() const noexcept {
+  return state_ == State::contiguous && static_cast<size_t>(views_count_) == size();
+}
+
+
+template <typename Backend>
+void TensorVector<Backend>::SetContiguous(bool contiguous) {
+  if (contiguous) {
+    state_ = State::contiguous;
+  } else {
+    state_ = State::noncontiguous;
+  }
+}
+
+
+template <typename Backend>
+void TensorVector<Backend>::Reset() {
+  tensors_.clear();
+  curr_tensors_size_ = 0;
+  type_ = {};
+  if (IsContiguous()) {
+    views_count_ = 0;
+    tl_->Reset();
+  }
+}
+
+
+template <typename Backend>
+template <typename SrcBackend>
+void TensorVector<Backend>::Copy(const TensorList<SrcBackend> &in_tl, cudaStream_t stream) {
+  SetContiguous(true);
+  tl_->Copy(in_tl, stream);
+
+  resize_tensors(tl_->ntensor());
+  views_count_ = 0;
+  UpdateViews();
+}
+
+
+template <typename Backend>
+template <typename SrcBackend>
+void TensorVector<Backend>::Copy(const TensorVector<SrcBackend> &in_tv, cudaStream_t stream) {
+  SetContiguous(true);
+  tl_->Copy(in_tv, stream);
+
+  resize_tensors(tl_->ntensor());
+  views_count_ = 0;
+  UpdateViews();
+}
+
+
+template <typename Backend>
+void TensorVector<Backend>::ShareData(TensorList<Backend> *in_tl) {
+  DALI_ENFORCE(in_tl != nullptr, "Input TensorList is nullptr");
+  SetContiguous(true);
+  pinned_ = in_tl->is_pinned();
+  tl_->ShareData(in_tl);
+
+  resize_tensors(in_tl->ntensor());
+  views_count_ = 0;
+  UpdateViews();
+}
+
+
+template <typename Backend>
+void TensorVector<Backend>::ShareWith(TensorList<Backend> *in_tl) const {
+  DALI_ENFORCE(in_tl != nullptr, "Input TensorList is nullptr");
+  if (IsContiguous()) {
+    in_tl->ShareData(tl_.get());
+    for (size_t i = 0; i < size(); ++i) {
+      in_tl->SetMeta(i, this->GetMeta(i));
+    }
+    in_tl->SetLayout(this->GetLayout());
+  } else {
+    DALI_FAIL("Cannot share non contiguous TensorVector with TensorList");
+  }
+}
+
+
+template <typename Backend>
+void TensorVector<Backend>::ShareData(TensorVector<Backend> *tv) {
+  DALI_ENFORCE(tv != nullptr, "Input TensorVector is nullptr");
+  state_ = tv->state_;
```

Comment:
```suggestion
  type_ = tv->type_;
  state_ = tv->state_;
```
Your change in `type()` has uncovered that setting `type_` is missing here (and probably in some other places - check Copy and ShareData).

Reply:
Done